src/shared/__tests__/errors.unit.test.ts

```diff
@@ -0,0 +1,92 @@
+import ERRORS from "../errors";
+
+describe("ERRORS", () => {
+  it("returns valid errors requested via dot queries", () => {
+    const err = ERRORS.__TEST.__NESTED;
+    expect(err).toBe("$");
+  });
+
+  it("returns valid errors requested via array queries", () => {
+    const err = ERRORS["__TEST.__NESTED"];
+    expect(err).toBe("$");
+  });
+
+  it("returns cached error messages correctly", () => {
+    const testErr = ERRORS.__TESTING;
+    expect(ERRORS.__TESTING).toBe(testErr);
+  });
+
+  it("returns functional errors", () => {
+    const err = ERRORS.__TEST.__FUNC;
+    expect(err.test).toBe("test");
+    expect(err.test_string_with_spaces).toBe("test string with spaces");
+  });
+
+  it("returns error messages with spaces verbatim", () => {
+    const err = "a b c";
+    expect(ERRORS[err]).toBe(err);
+  });
+
+  it("throws an error upon attempting to access an invalid error requested via dot queries", () => {
+    try {
+      ERRORS.__TEST.__INVALID_ERROR_MESSAGE;
+      expect("Error was not thrown by ErrorsProxy").toBe(false);
+    } catch (e) {
+      expect(e).toBeInstanceOf(ReferenceError);
+      expect(e.message).not.toMatch("Error was not thrown by ErrorsProxy");
+    }
+  });
+
+  it("throws an error upon attempting to access an invalid error requested via array queries", () => {
+    try {
+      ERRORS["__TEST.__INVALID_ERROR_MESSAGE"];
+      expect("Error was not thrown by ErrorsProxy").toBe(false);
+    } catch (e) {
+      expect(e).toBeInstanceOf(ReferenceError);
+      expect(e.message).not.toMatch("Error was not thrown by ErrorsProxy");
+    }
+  });
+
+  it("rejects modifying any top-level errors", () => {
+    try {
+      ERRORS.__TESTING = "Bad!";
+      expect("Error was not thrown by ErrorsProxy").toBe(false);
+    } catch (e) {
+      expect(e.message).not.toMatch("Error was not thrown by ErrorsProxy");
+    }
+  });
+
+  it("rejects modifying any nested errors requested via dot queries", () => {
+    try {
+      ERRORS.__TEST.__NEST_1.__NEST_2 = "Bad!";
+      expect("Error was not thrown by ErrorsProxy").toBe(false);
+    } catch (e) {
+      expect(e.message).not.toMatch("Error was not thrown by ErrorsProxy");
+    }
+  });
+
+  it("rejects adding additional errors to the top level", () => {
+    try {
+      ERRORS.__ASDFUHFAI = "?";
+      expect("Error was not thrown by ErrorsProxy").toBe(false);
+    } catch (e) {
+      expect(e.message.includes("Error was not thrown by ErrorsProxy")).toBe(false);
+    }
+  });
+
+  it("rejects adding additional errors to nested levels", () => {
+    try {
+      ERRORS.__TEST.__ASDFUHFAI = "?";
+      expect("Error was not thrown by ErrorsProxy").toBe(false);
+    } catch (e) {
+      expect(e.message.includes("Error was not thrown by ErrorsProxy")).toBe(false);
+    }
+  });
+
+  describe("USER.REQUIRED", () => {
```

Comment:
We'll probably end up using the required form in more places than the user form, so we should move it to the top-level.

Reply:
The way I structured this test suite (and all my test suites) is that top-level test groups all correspond to a module export. We then test functionality related to that export within the test group. If a method or part of an export requires more granular testing (e.g. here with `USER.REQUIRED`), we declare a nested test group. This way, the organization of test suites follow the organization of their corresponding modules.

Reply:
I probably worded that weird. What I meant is that in the error map, we should move `USER.REQUIRED` into just `REQUIRED`, seeing as the text is pretty generic and doesn't reference the user form at all.

Reply:
Ahh, I see what you mean. Moved and deprecated in 02c0bea.